File: python/slhdsa/slhdsa.py
"""SLH-DSA top-level keygen, sign, verify per FIPS 205, Section 9."""

import os
import math

from slhdsa.params import SLHDSAParams
from slhdsa.address import ADRS, FORS_TREE
from slhdsa.hash_funcs import get_suite
from slhdsa.hypertree import ht_sign, ht_verify
from slhdsa.fors import fors_sign, fors_pk_from_sig
from slhdsa.xmss import xmss_node
from slhdsa.utils import to_int


def keygen(params: SLHDSAParams) -> tuple:
    """SLH-DSA key generation (Algorithm 17).

    Returns:
        (pk, sk) where pk = PK.seed || PK.root and
        sk = SK.seed || SK.prf || PK.seed || PK.root.
    """
    sk_seed = os.urandom(params.n)
    sk_prf = os.urandom(params.n)
    pk_seed = os.urandom(params.n)

    adrs = ADRS()
    adrs.set_layer_address(params.d - 1)
    adrs.set_tree_address(0)

    pk_root = xmss_node(params, sk_seed, 0, params.hp, pk_seed, adrs)

    pk = pk_seed + pk_root
    sk = sk_seed + sk_prf + pk_seed + pk_root
    return pk, sk


def sign(sk: bytes, m: bytes, params: SLHDSAParams,
         randomize: bool = True) -> bytes:
    """SLH-DSA signature generation (Algorithm 18).

    Args:
        sk: Secret key.
        m: Message to sign.
        params: Parameter set.
        randomize: If True, use randomized signing.

    Returns:
        SLH-DSA signature.
    """
    suite = get_suite(params.hash_type)
    n = params.n

    # Parse secret key
    sk_seed = sk[0:n]
    sk_prf = sk[n:2*n]
    pk_seed = sk[2*n:3*n]
    pk_root = sk[3*n:4*n]

    # Generate randomizer
    if randomize:
        opt_rand = os.urandom(n)
    else:
        opt_rand = pk_seed

    # R = PRF_msg(SK.prf, opt_rand, M)
    r = suite.PRF_msg(sk_prf, opt_rand, m, n)

    # Generate message digest
    sig = r  # First part of signature

    digest = suite.H_msg(r, pk_seed, pk_root, m, params.m)

    # Split digest into md, idx_tree, idx_leaf
    # md: ceil(k*a / 8) bytes
    # idx_tree: ceil((h - h/d) / 8) bytes
    # idx_leaf: ceil((h/d) / 8) bytes
    md_len = math.ceil(params.k * params.a / 8)
    tree_bits = params.h - params.hp
    tree_len = math.ceil(tree_bits / 8)
    leaf_len = math.ceil(params.hp / 8)

    md = digest[0:md_len]
    idx_tree_bytes = digest[md_len:md_len + tree_len]
    idx_leaf_bytes = digest[md_len + tree_len:md_len + tree_len + leaf_len]

    idx_tree = to_int(idx_tree_bytes) % (1 << tree_bits)
    idx_leaf = to_int(idx_leaf_bytes) % (1 << params.hp)

    # FORS signature
    adrs = ADRS()
    adrs.set_layer_address(0)
    adrs.set_tree_address(idx_tree)
    adrs.set_type(FORS_TREE)
    adrs.set_key_pair_address(idx_leaf)

    sig_fors = fors_sign(params, md, sk_seed, pk_seed, adrs)
    sig += sig_fors

    # Get FORS public key for HT signing
    pk_fors = fors_pk_from_sig(params, sig_fors, md, pk_seed, adrs)

    # Hypertree signature
    sig_ht = ht_sign(params, pk_fors, sk_seed, pk_seed, idx_tree, idx_leaf)
    sig += sig_ht

    return sig


def verify(pk: bytes, m: bytes, sig: bytes, params: SLHDSAParams) -> bool:
    """SLH-DSA signature verification (Algorithm 19).

    Args:
        pk: Public key.
        m: Message.
        sig: Signature.
        params: Parameter set.

    Returns:
        True if signature is valid.
    """
    suite = get_suite(params.hash_type)
    n = params.n

    if len(sig) != params.sig_size:
        return False
    if len(pk) != params.pk_size:
        return False

    # Parse public key
    pk_seed = pk[0:n]
    pk_root = pk[n:2*n]

    # Parse signature
    r = sig[0:n]
    fors_sig_size = params.k * (params.a + 1) * n
    sig_fors = sig[n:n + fors_sig_size]
    sig_ht = sig[n + fors_sig_size:]

    # Recompute message digest
    digest = suite.H_msg(r, pk_seed, pk_root, m, params.m)

    md_len = math.ceil(params.k * params.a / 8)
    tree_bits = params.h - params.hp
    tree_len = math.ceil(tree_bits / 8)
    leaf_len = math.ceil(params.hp / 8)

    md = digest[0:md_len]
    idx_tree_bytes = digest[md_len:md_len + tree_len]
    idx_leaf_bytes = digest[md_len + tree_len:md_len + tree_len + leaf_len]

    idx_tree = to_int(idx_tree_bytes) % (1 << tree_bits)
    idx_leaf = to_int(idx_leaf_bytes) % (1 << params.hp)

    # Recompute FORS public key
    adrs = ADRS()
    adrs.set_layer_address(0)
    adrs.set_tree_address(idx_tree)
    adrs.set_type(FORS_TREE)
    adrs.set_key_pair_address(idx_leaf)

    pk_fors = fors_pk_from_sig(params, sig_fors, md, pk_seed, adrs)

    # Verify hypertree signature
    return ht_verify(params, pk_fors, sig_ht, pk_seed, idx_tree, idx_leaf, pk_root)

File: js/src/slhdsa/slhdsa.js
/**
 * SLH-DSA Top-Level Operations (FIPS 205)
 * Algorithms 18-20: keyGen, sign, verify
 */

import { randomBytes } from 'node:crypto';
import { ADRS, ADDR_TYPE } from './address.js';
import { createHashSuite } from './hash.js';
import { xmssNode } from './xmss.js';
import { forsSign, forsPkFromSig } from './fors.js';
import { htSign, htVerify } from './hypertree.js';
import { toInt, toBigInt, base2b, concat } from './utils.js';

/**
 * slhKeyGen(params) - Algorithm 18
 * Generate SLH-DSA key pair
 * Returns { sk, pk } where:
 *   sk = skSeed || skPrf || pkSeed || pkRoot (4*n bytes)
 *   pk = pkSeed || pkRoot (2*n bytes)
 */
function slhKeyGen(params) {
  const n = params.n;
  const hashSuite = createHashSuite(params);

  // Generate random seeds
  const skSeed = new Uint8Array(randomBytes(n));
  const skPrf = new Uint8Array(randomBytes(n));
  const pkSeed = new Uint8Array(randomBytes(n));

  // Compute root of the top XMSS tree in the hypertree
  const adrs = new ADRS();
  adrs.setLayerAddress(params.d - 1);
  const pkRoot = xmssNode(skSeed, pkSeed, 0, params.hp, adrs, params, hashSuite);

  const sk = concat(skSeed, skPrf, pkSeed, pkRoot);
  const pk = concat(pkSeed, pkRoot);

  return { sk, pk };
}

/**
 * slhSign(msg, sk, params, optRand) - Algorithm 19
 * Sign a message with SLH-DSA
 *
 * msg: Uint8Array message
 * sk:  Uint8Array secret key (4*n bytes)
 * optRand: optional randomness (n bytes), if null uses pkSeed (deterministic)
 *
 * Returns signature as Uint8Array
 */
function slhSign(msg, sk, params, optRand) {
  const n = params.n;
  const hashSuite = createHashSuite(params);

  // Parse secret key
  const skSeed = sk.slice(0, n);
  const skPrf = sk.slice(n, 2 * n);
  const pkSeed = sk.slice(2 * n, 3 * n);
  const pkRoot = sk.slice(3 * n, 4 * n);

  // Generate randomizer
  if (!optRand) {
    optRand = pkSeed; // deterministic variant
  }
  const R = hashSuite.PRFMsg(skPrf, optRand, msg);

  // Compute message digest
  const digest = hashSuite.HMsg(R, pkSeed, pkRoot, msg);

  // Parse digest into FORS message, tree index, leaf index
  const { mdBytes, treeIdxBytes, leafIdxBytes, k, a, forsMsgBits, treeIdxBits, leafIdxBits } = params;

  const mdBuf = digest.slice(0, mdBytes);
  const treeIdxBuf = digest.slice(mdBytes, mdBytes + treeIdxBytes);
  const leafIdxBuf = digest.slice(mdBytes + treeIdxBytes);

  // FORS message: k values, each a bits
  const md = base2b(mdBuf, a, k);

  // Tree and leaf indices
  let idxTree = toBigInt(treeIdxBuf, treeIdxBytes);
  // Mask to treeIdxBits
  const treeMask = (1n << BigInt(treeIdxBits)) - 1n;
  idxTree &= treeMask;

  let idxLeaf = toInt(leafIdxBuf, leafIdxBytes);
  // Mask to leafIdxBits
  idxLeaf &= (1 << leafIdxBits) - 1;

  // FORS signature
  const adrs = new ADRS();
  adrs.setTreeAddress(idxTree);
  adrs.setType(ADDR_TYPE.FORS_TREE);
  adrs.setKeyPairAddress(idxLeaf);

  const sigFors = forsSign(md, skSeed, pkSeed, adrs, params, hashSuite);

  // Compute FORS public key
  const pkFors = forsPkFromSig(sigFors, md, pkSeed, adrs, params, hashSuite);

  // Hypertree signature on FORS public key
  const sigHt = htSign(pkFors, skSeed, pkSeed, idxTree, idxLeaf, params, hashSuite);

  // Return R || sigFors || sigHt
  return concat(R, sigFors, sigHt);
}

/**
 * slhVerify(msg, sig, pk, params) - Algorithm 20
 * Verify an SLH-DSA signature
 *
 * Returns true if valid, false otherwise
 */
function slhVerify(msg, sig, pk, params) {
  const n = params.n;
  const hashSuite = createHashSuite(params);

  // Check signature length
  if (sig.length !== params.sig) {
    return false;
  }

  // Parse public key
  const pkSeed = pk.slice(0, n);
  const pkRoot = pk.slice(n, 2 * n);

  // Parse signature
  const R = sig.slice(0, n);
  const forsSigLen = params.k * (1 + params.a) * n;
  const sigFors = sig.slice(n, n + forsSigLen);
  const sigHt = sig.slice(n + forsSigLen);

  // Compute message digest
  const digest = hashSuite.HMsg(R, pkSeed, pkRoot, msg);

  const { mdBytes, treeIdxBytes, leafIdxBytes, k, a, treeIdxBits, leafIdxBits } = params;

  const mdBuf = digest.slice(0, mdBytes);
  const treeIdxBuf = digest.slice(mdBytes, mdBytes + treeIdxBytes);
  const leafIdxBuf = digest.slice(mdBytes + treeIdxBytes);

  const md = base2b(mdBuf, a, k);

  let idxTree = toBigInt(treeIdxBuf, treeIdxBytes);
  const treeMask = (1n << BigInt(treeIdxBits)) - 1n;
  idxTree &= treeMask;

  let idxLeaf = toInt(leafIdxBuf, leafIdxBytes);
  idxLeaf &= (1 << leafIdxBits) - 1;

  // Compute FORS public key from signature
  const adrs = new ADRS();
  adrs.setTreeAddress(idxTree);
  adrs.setType(ADDR_TYPE.FORS_TREE);
  adrs.setKeyPairAddress(idxLeaf);

  const pkFors = forsPkFromSig(sigFors, md, pkSeed, adrs, params, hashSuite);

  // Verify hypertree signature
  return htVerify(pkFors, sigHt, pkSeed, idxTree, idxLeaf, pkRoot, params, hashSuite);
}

export { slhKeyGen, slhSign, slhVerify };
